blingfire/src/lib.rs

```diff
@@ -1,100 +1,71 @@
 mod errors;
 
 use blingfire_sys::{TextToSentences as text_to_sentences_ffi, TextToWords as text_to_words_ffi};
-use failchain::{bail, ResultExt};
-use std::{convert::TryInto, mem};
+use failchain::ensure;
+use std::convert::TryInto;
+use std::i32;
+use std::os::raw::{c_char, c_int};
 
 pub use crate::errors::{ErrorKind, Result};
 
 #[inline]
 pub fn text_to_words(source: &str, destination: &mut String) -> Result<()> {
-    tokenize(
-        |raw_source, raw_source_len, raw_destination, raw_destination_capacity| unsafe {
-            text_to_words_ffi(
-                raw_source,
-                raw_source_len,
-                raw_destination,
-                raw_destination_capacity,
-            )
-        },
-        source,
-        destination,
-    )
+    tokenize(text_to_words_ffi, source, destination)
 }
 
 #[inline]
 pub fn text_to_sentences(source: &str, destination: &mut String) -> Result<()> {
-    tokenize(
-        |raw_source, raw_source_len, raw_destination, raw_destination_capacity| unsafe {
-            text_to_sentences_ffi(
-                raw_source,
-                raw_source_len,
-                raw_destination,
-                raw_destination_capacity,
-            )
-        },
-        source,
-        destination,
-    )
+    tokenize(text_to_sentences_ffi, source, destination)
 }
 
+type Tokenizer = unsafe extern "C" fn(*const c_char, c_int, *mut c_char, c_int) -> c_int;
+
+// The maximum length is when the destination length can't fit in the maximum blingfire array. Worst
+// case, the source is tokenized by adding a space after every character + a null character at the
+// end.
+//
+// The maximum array size is defined in blingfire-sys/BlingFire/blingfireclient.library/inc/FALimits.h
+const MAX_SOURCE_LENGTH: usize = 1_000_000_000 / 2 - 2;
+
 #[inline]
-fn tokenize<Tokenizer>(tokenizer: Tokenizer, source: &str, destination: &mut String) -> Result<()>
+fn tokenize(tokenizer: Tokenizer, source: &str, destination: &mut String) -> Result<()>
```

Comment:
As I mentioned, I think this change may have performance implications, you're passing in a fn pointer now, if it affects inlining I think it should stay generic